Code Review

efrapp opened the pull request.

efrapp committed:
Design models
- Create database diagram for models' association
- Create User model to send and accept friends' request

efrapp committed:
Update README.md

rvvergara commented on Gemfile:
```diff
@@ -36,6 +36,8 @@ gem 'jbuilder', '~> 2.5'
 # Reduces boot times through caching; required in config/boot.rb
 gem 'bootsnap', '>= 1.1.0', require: false
 
+gem 'hirb'
```
You might wanna put this gem into the `development, test` group. Also I suggest using `pry-rails` in addition to `hirb`. Makes your console display look so much better :)

rvvergara reviewed (approved):
Hi @efrapp and @userman95,

Congratulations on coming this far! This is your final Rails project (at least from the program's standpoint). So we'll be making sure you build a very nice app.

## Suggested Testing Framework
First I'd like to say to use RSpec for testing. In my experience, it's much easier to write specs/tests and the test data is less brittle as it uses `factories` instead of `fixtures`.

## ERD

I think you should fill up your `Post`, `Like` and `Comment` as they don't depict the `user_id`. We should not be able to create `posts`, `comments` or `likes` without them attached to a `user`.

For other matters that may arise from your data architecture we can tackle once you have created your models. Please be sure to test each of your models for validations and associations. And before even starting to create controllers or views let's first make sure all your models are stable. For this, the `rails console` is your best friend. You will need to try out all the features of your models there. 

Ok, I believe you are ready now to move on and build your models. So:

# :+1:

orestis-kaplan committed:
Adding models and devise gem logic
    -Adding associations for comment,like and post models
    -Adding devise gem ,some html warnings and root route to signup
    page

efrapp committed:
Adding testing for model User

orestis-kaplan committed:
Tests for models
    - Comment model
    - Post model
    - Like model

efrapp committed:
User testing
- creating tests for user's field validations and associations
- Adding fixtures for users

efrapp committed:
Solving merge conflicts from remote repo

stickler-ci committed:
Adding .stickler.yml

efrapp committed:
Add stickler file

efrapp committed:
Update .stickler.yml

Adding config for Ruby checking

efrapp committed:
Merge pull request #3 from efrapp/add-stickler-config

Add stickler config

efrapp committed:
Remove Stickler file from repo

efrapp committed:
Merge branch 'develop' of github.com:efrapp/facebook into develop

efrapp committed:
Remove Stickler file from repo

efrapp commented:
Sending new code review for Models

efrapp committed:
Move Hirb gem to development group on GemFile

rvvergara commented on db/schema.rb:
```diff
@@ -0,0 +1,71 @@
+  # These are extensions that must be enabled in order to support this database
+  enable_extension "plpgsql"
+
+  create_table "comments", force: :cascade do |t|
```
Why aren't we referencing `user` here in comments? It will be a problem in your implementation later on without such reference.

See #4

efrapp replied:
User reference added to comments.

rvvergara commented on db/schema.rb:
```diff
@@ -0,0 +1,71 @@
+    t.index ["sender_id"], name: "index_friendships_on_sender_id"
+  end
+
+  create_table "likes", force: :cascade do |t|
```
Like in `comments`, why are we not referencing `user` here? 

See #4

efrapp replied:
User reference added to likes.

rvvergara commented on app/models/user.rb:
```diff
@@ -0,0 +1,26 @@
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable
+         
+  has_many :requests, 
```
Why did we use `:requests` as association name for 'Friendship`? Why not `:friends`? This is a little confusing

efrapp replied:
Ok Ryan, I changed `:requests` to `:sent_requests`.

rvvergara commented on app/models/user.rb:
```diff
@@ -0,0 +1,26 @@
+           class_name: "Friendship",
+           foreign_key: :sender_id,
+           dependent: :destroy
+  has_many :invitations,
```
Same comment as for `:requests` above.

I recognize what you wanna do here. You want to distinguish a user that initiated the friendship vs the user that received the friend invitation. So I would stick to simple `sent_requests` and `received_requests` to make this more explicit.

efrapp replied:
Same here. I changed `invitations` to `received_requests`.

rvvergara commented on app/models/user.rb:
```diff
@@ -0,0 +1,26 @@
+           dependent: :destroy
+
+  # Invitations pending to accept
+  has_many :invitations_from, through: :invitations, source: :sender
```
How can we, for instance, know `efrain` 's list of friends (which means those that have accepted his friend requests or those whose friend requests efrain already accepted)?

We probably need a method that will let us list all of a user's set of friends.

efrapp replied:
If we want all the friends that accepted the `efrain`'s request we can do this:
```
efrain.sent_requests.where(accepted: true)
```
And if we want the friend requests `efrain` already accepted we can use:
```
efrain.received_requests.where(accepted: true)
```
I also made two test cases for that.

rvvergara commented on app/models/user.rb:
```diff
@@ -0,0 +1,26 @@
+  has_many :requests_to, through: :requests, source: :receiver
+
+  has_many :posts, dependent: :destroy
+  has_many :likes
```
Are we not also destroying a `user` 's `comments` and `likes` should the `user` 's account be destroyed?

efrapp replied:
Ok, we added the `dependent` option for both, `likes` and `comments`.

rvvergara commented on app/models/comment.rb:
```diff
@@ -0,0 +1,5 @@
+class Comment < ApplicationRecord
```
See issue #4

efrapp replied:
The `belongs_to` association is already placed in the model.

rvvergara commented on app/models/like.rb:
```diff
@@ -0,0 +1,3 @@
+class Like < ApplicationRecord
```
See issue #4

efrapp replied:
`belongs_to` added to Like model too.

rvvergara reviewed (changes requested):
Hi @efrapp and @userman95,

Great leap forward in building your models! Though there are some work we need to do. Please see my comments in the respective model files and issue #4.

In regards to testing, please use RSpec as your testing framework. Minitest is good but RSpec is more robust, less brittle and more widely used. Read [this](https://medium.com/back-to-rails/setting-up-rspec-testing-for-rails-5-28be8431f845) for more information on how to set up RSPec for your project. Use `factory_bot_rails` instead of the `factory_girl` gem mentioned in that article.

orestis-kaplan committed:
Add the omniauth for facebook ,get it working and add references from comments and likes to users

orestis-kaplan committed:
Update Gemfile.lock for Heroku

orestis-kaplan committed:
chenges for heroku

orestis-kaplan committed:
adding heroku path to devise.rb

orestis-kaplan committed:
adding heroku path to devise.rb

orestis-kaplan committed:
adding heroku path to devise.rb

orestis-kaplan committed:
adding heroku path to devise.rb

orestis-kaplan committed:
make chenges for heroku deploy

orestis-kaplan committed:
make chenges for heroku deploy

orestis-kaplan committed:
make chenges for heroku deploy

orestis-kaplan committed:
make chenges for heroku deploy

orestis-kaplan committed:
make chenges for heroku deploy

orestis-kaplan committed:
make chenges for heroku deploy

orestis-kaplan committed:
make chenges for heroku deploy

orestis-kaplan committed:
make chenges for heroku deploy

orestis-kaplan committed:
adding further migrations for trackable

orestis-kaplan committed:
Changed omniauth to localhost for testing and fix model logic

efrapp committed:
Add user tests to check accepted and pending friendship requests from both sides

efrapp committed:
Merge branch 'develop' of github.com:efrapp/facebook into develop
- Adding user foreign key for comments and likes

efrapp committed:
Update test for likes and comments

efrapp commented:
Changes finished for second code review on models

orestis-kaplan committed:
Send friend request to user
	-send_request action
	-routes for send_request
	-partial for index action
Begin basic html and css lookup

rvvergara commented on app/models/user.rb:
```diff
@@ -0,0 +1,45 @@
+         :recoverable, :rememberable, :trackable , :validatable,
+         :omniauthable, :omniauth_providers => [:facebook]
+
+  has_many :sent_requests,
```
While you've addressed my previous question regarding having a method that will list all `user`'s friends, I still highly encourage you to have such a method instead of having to write out a long query everytime you need a user's list of friends. This will make your controllers and views later on easier to build. 

Also this makes your data easier to reason about.

I am applying here the principle of **fat-model thin-controllers/views**. I am sure that later in your views you would want to list a user's friends, mutual friends with another user, etc. This will necessitate you to query over and over again the database (using the `where` statement you previously mentioned) just to do this. That will make your logic transfer more and more to your controllers, which isn't a good practice.

rvvergara commented on app/models/friendship.rb:
```diff
@@ -0,0 +1,4 @@
+class Friendship < ApplicationRecord
```
Please see #5

rvvergara commented on test/models/user_test.rb:
```diff
@@ -0,0 +1,127 @@
+require 'test_helper'
+
+class UserTest < ActiveSupport::TestCase
```
I highly highly encourage you guys to use RSPec for testing. For one it will make your specs easier to debug and easier to maintain. Secondly, RSpec's syntax is much readable. Third, fixtures, which Minitest uses, is brittle (easily breaks) than RSpec's factories. A professional benefit is that you'll find more companies using RSpec more than Minitest. Also RSpec has closer resemblance with Jest (which you will use when you get to the JavaScript sections). I hope these benefits convince you why it's great to convert into RSpec. :)

rvvergara reviewed (changes requested):
Hi @efrapp and @userman95,

Pls see my latest reviews. I would normally not want to get in your way on how you design your data architecture. But I think it would be better if you have a separate model for recording FriendRequests rather than stuffing it all up under `Friendship`. Here are my two cents worth:

1. As a data maintainer you will have an easier time handling massive records later on. Stuffing all friend requests as if they're already friendships will make it confusing later on. This is even if you have the `accepted` column. Do I say `mike` has `ryto` as a friend even if `ryto` has not accepted mike yet as a friend?

2. Related to my comment in `user.rb` and 1) above, it makes your db query more complicated.

But then again it is entirely up to you on how to implement this. I just wanted to share a better approach in my perspective. 

I also suggest you make full use of your models as the home to your advanced database querying. Controllers should just, for the most part, do simple `find` and/or `find_by` queries. This is in relation to the **fat-model thin-controller** principle. I haven't checked any controllers yet in this project but based on my current and previous review I guess if I don't see much queries in your model (especially with the user model) then most of it will go to the controllers and/or views (which is bad).

efrapp committed:
Add pending request feature (in progress)

orestis-kaplan committed:
updating the models.Adding a FriendRequest model

orestis-kaplan committed:
Installing Rspec and some supportive gems

orestis-kaplan committed:
changed factory girl

efrapp committed:
Merge branch 'develop' of github.com:efrapp/facebook into develop

efrapp committed:
Add test for users using RSpec (in progress)

efrapp committed:
Remove let from association test on user

orestis-kaplan committed:
Building the new structure for models
	-User model
	-Friendship model
	-Friend Request model
	-Like model
	-Comment model
	-Post model
Build tests for all the above models with rspec.
Build functions for validation in :
	-Friendship model
	-Friend Request model

orestis-kaplan committed:
Fix conficts

orestis-kaplan committed:
Adding test to user model

orestis-kaplan committed:
Styles for header and a couple of functions in models

rvvergara reviewed (changes requested):
Hi @efrapp and @userman95,

Pls see my latest review :) Please also check #6. Ok while I reviewed your `users_controller.rb` file, I assume that this review is still a continuation of our reviews solely for models. I hope the references and code snippets I provided will be helpful.

userman95 commented on app/models/user.rb:
```diff
@@ -0,0 +1,46 @@
+         :recoverable, :rememberable, :trackable , :validatable,
+         :omniauthable, :omniauth_providers => [:facebook]
+
+  has_many :friend_requests, dependent: :destroy
```
ok . So if mike want to send to ryto a friend request it goes like this: ryto.friend_requests.create(friend: mike) .And ryto can see his friend requests like this: ryto.friend_requests.

orestis-kaplan committed:
Changing friendship model

efrapp committed:
Testing Models
- Adding FactoryBot gem to use factories
- Adding association tests for user using FactoryBot, let and the new Friendship flow

orestis-kaplan committed:
nothing at all

efrapp committed:
Testing User
- Add method to return the friend requests between two users
- Add method to return the users involved in a friend request
- Add testing for user associations

efrapp committed:
Merge branch 'develop' of github.com:efrapp/facebook into develop

efrapp committed:
Add association tests for User using factories

orestis-kaplan committed:
Building the controllers and the views.

efrapp committed:
Models' validation
- Update User validation using let and factories
- Update FriendRequest validation using let and factories
- Update Friendship validation using let and factories

efrapp committed:
Tests for Models
- Updating tests for Post model
- Updating tests for Like model

efrapp committed:
Update tests for Comment model

efrapp committed:
Merge branch 'develop' of github.com:efrapp/facebook into develop

orestis-kaplan committed:
Commiting the basic structure of facebook and implementing AJAX request for likes with buttons

orestis-kaplan committed:
Merge branch 'develop' of github.com:efrapp/facebook into develop

orestis-kaplan committed:
fixing likes button functionality

orestis-kaplan committed:
fixing likes button functionality

rvvergara commented on app/models/user.rb:
```diff
@@ -0,0 +1,71 @@
+  has_many :pending_acceptances, through: :received_requests,
+    source: :user, dependent: :destroy
+
+  has_many :active_friendships,class_name: "Friendship",
```
Great job at this guys! :)

rvvergara commented on app/models/friendship.rb:
```diff
@@ -0,0 +1,22 @@
+
+  private
+  
+    def friendship_uniqueness
```
Guys I don't want to bog you down on this. But if you're interested, you can go further and enforce this uniqueness at the database level on top of this validation you are already doing. If you like to do this just reach out to me on slack :)

efrapp replied:
Ryan, I added the code to enforce uniqueness in the database side as you suggested but I had to use `before_validation` callback to give priority to the `check_duplicates` method, any other callback will run after validations so we can't get the desired result. Let me know your thoughts, please.

rvvergara commented on app/models/friend_request.rb:
```diff
@@ -0,0 +1,42 @@
+  validate :uniqueness_of_friend_request
+  validate :exclude_self_friend_request
+
+  def accept
```
Probably we don't need this method anymore. We can simply use the CRUD actions for Friendship to create the associated friendship record :)

efrapp replied:
Method removed. Also, the specs for models were updated to remove the `accept` method.

rvvergara commented on app/models/friend_request.rb:
```diff
@@ -0,0 +1,42 @@
+    destroy
+  end
+
+  def decline
```
Likewise, perhaps we don't need this special method anymore. We can simply use the `destroy` directly when a user clicks on the `Decline Request` button

efrapp replied:
Method removed. Also, the specs for models were updated.

rvvergara commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,78 @@
+  let(:friend){ build(:user, name: "Orestis", email: "o@mail.c") }
+  let(:requested_users){ create_list(:user, 10) }
+
+  context 'validation tests' do
```
Guys, i suggest the following organization for your specs to make it more readable:

```ruby
describe "validations" do
  context "complete user details" do
    it "is valid" do
      # expectations here
    end
  end
  
   context "name absent" do
      it "is invalid" do
        # expectations here
      end
   end
end
```

efrapp replied:
We changed the `context` block with `describe` but we just used `context` in some sections because it wasn't needed to use it in all the specs.

rvvergara reviewed (changes requested):
Hi @efrapp and @userman95,

Great job after our conversation! :) Pls see my updated review here. Your models are almost good to go and I suggest for the next review you already include some work on the controllers already. Most of my comments here are just for refinement of the friendship/friend_request records and for the specs.

efrapp committed:
Use describe instead of context as the top level block for specs

efrapp committed:
Enforce uniqueness in the database side

efrapp committed:
Models and Specs
- Create callback to delete friend request after friendship is created
- Update specs to support the new callback

efrapp committed:
Improve expect for user test

efrapp commented:
New code review sent for models

efrapp committed:
Refactoring
- Fix routes for posts and users
- Add error and notification messages
- Update render for post form
- Update create action for post controller
Post feature
- Upload picture for post

orestis-kaplan committed:
adding logic for likes and controller , building the views of them

orestis-kaplan committed:
implement searching

orestis-kaplan committed:
building the friend requests, manage friendships,adding buttonsand add js functionalities

orestis-kaplan committed:
Adding images everywhere that are needed , fixing styles on friend request

orestis-kaplan committed:
Fixing style issues

efrapp committed:
Refactoring
- users, posts, likes, comments controller
- create separate partials for like and comment buttons
- fix toggle for comment button

efrapp committed:
Merge branch 'develop' of github.com:efrapp/facebook into develop
- Add public/upload folder to .gitignore file

efrapp committed:
Remove uploads folder from git tracking

orestis-kaplan committed:
adding changes for the resize of pictures and fixing styles in the header

efrapp committed:
Refactoring
- refactor for friend requests and friendships controllers
- create scope in user model

efrapp committed:
Merge branch 'develop' of github.com:efrapp/facebook into develop

efrapp committed:
Fix error on click comment event

efrapp committed:
move delete_friend_request action to a callback

rvvergara commented on app/controllers/friendships_controller.rb:
```diff
@@ -0,0 +1,24 @@
+
+  def create
+    @friend = User.find_by(id: params[:friend_id])
+    @friendship = current_user.active_friendships.build(friend: @friend)
```
There isn't any major issue with this line, but I would like to suggest to make the attributes of `Friendship` consistent with `FriendRequest`. What I mean is that for example in this line of code, we are making the user that received the `FriendRequest` as the `active` friend. I suggest let's make the receiver of the `FriendRequest` the `friend_id` and the sender of the `FriendRequest` as the `user_id` in the `Friendship` record. If that's confusing, in code:

```ruby
@friendship = current_user.passive_friendships.build(user: @friend)
```
I think reflects more appropriately what is happening. Though again, if you sill prefer what you put here, it's all good :)

rvvergara commented on app/controllers/posts_controller.rb:
```diff
@@ -0,0 +1,32 @@
+
+  def create
+    @post = current_user.posts.build(post_params)
+    if @post.save
```
This is a great idea of redirecting a user back to the page where he/she created the post. However, normally we'd like the `if else` block to be all about either successfully creating a post or not. To redirect a user back to the place where he/she has created a post I suggest you investigate using `redirect_to_session`. I think this was used in the Rails Tutorial Book

rvvergara reviewed (changes requested):
Hi @efrapp and @userman95,

Ok now we're on controllers and views. Congratulations for making it this far into the project and into the program!! :) You guys are awesome!

## User Interface

I like the part that I can search for users :) Great job on this!

I can make a post (both text and image), comment on it (and delete my comment), like and unlike it. => Awesome!! How about making posts deletable?

I can send a friend request and also accept it (and delete a friend). It would be nice if I could see a notification that I have a friend request.

### User show page

Haha im wondering what this image is and what it is for?

![image](https://user-images.githubusercontent.com/6015315/59985926-d5d75380-9666-11e9-9202-3337e4bd0825.png)

### Editing my profile

We have an `update` action on our `users_controller` but how do I go to the Edit user page?

## Let's write specs for features

Please also see my annotations to some controller files.

orestis-kaplan committed:
applying changes in controllers by suggestion

orestis-kaplan committed:
Merge branch 'develop' of github.com:efrapp/facebook into develop

orestis-kaplan committed:
applying changes in controllers by Ryan suggestion.Putting notification on friend request,add form for wallpaper image

orestis-kaplan committed:
Fixing friend requests from refactoring....

rvvergara reviewed (changes requested):
@efrapp and @userman95,

We're getting closer and closer to finishing this project. You guys are rocking it! :) 

Ok for your controllers, all is good now. 

For your views. I didn't dig too deep into each file as I trust that you are going to DRY them up with the appropriate use of partials. Just also watch out that you're not placing too much logic into your views (your views should be totally dumb and will just get their logic from the helpers/controllers). 

Some remaining tasks that need to be done:

Please see #7 

Also let's pls write `feature` specs.

efrapp committed:
Delete upload files from remote

orestis-kaplan committed:
Error handling for uploading image

efrapp committed:
Add feature spec for create a post

efrapp committed:
Create spec feature for delete post

orestis-kaplan committed:
fixing display error in show page

efrapp committed:
Create features for add and remove a like

efrapp committed:
Feature specs
- feature specs for Comments, Friend Requests and Friendships
- fix typo, replacing format: with formats: in some controllers
- fix log in problem on feature spec scenario for post

efrapp committed:
Merge branch 'develop' of github.com:efrapp/facebook into develop

efrapp committed:
Add missing migration for user wallpaper field

efrapp committed:
Update model specs for friendships and users with latest changes

rvvergara commented:
Hi @efrapp  and @userman95,

All is well with your app:

### Project Specification

1. Core features are included in the plan:
    - [x] users accounts
    - [x] friendships
    - [x] posts & comments

### Crucial things to make project work

- [x] The project has a descriptive README.md
- [x] Project should be done collaboratively and that the commit history reflects that

#### Database 
- [x] Use Postgresql for your database from the beginning (not sqlite3)

#### Models

##### User
- [x] Can send Friend Requests to other Users
- [x] Can accept the Friend Request
- [x] Must accept the Friend Request to become friends
- [x] Can create a Post
- [x] Can Like Posts
- [x] Can comment on Posts

##### Post
- [x] Has many comments
- [x] Has many likes

#### Controllers and Actions

##### Sessions
- [x] You must be logged in to watch anything but the login page
- [x] Login should use the devise gem
- [x] Allows logging in using a Facebook account through Omniauth

##### Navigation bar
- [x] Shows the number of received Friend Requests

##### User
- [x] Can create a Profile with a Photo

###### `:index`
- [x] Lists all users
- [x] Displays a button next to users to send a friend request
- [x] Displays if a friend request was sent
- [x] Displays if a user is already a friend
- [x] The current user can send Friend Requests to other Users

###### `:show`

- [x] Displays profile information
- [x] Displays the avatar
- [x] Displays the user's posts

##### Post

###### `:new`
- [x] The current user can create a new post

###### `:index`
- [x] Show recent posts from friends

###### `:show`
- [x] Displays the content
- [x] Displays the author
- [x] Displays the comments
- [x] Displays the likes
- [x] The current user can like the post
- [x] The current user comment on the post

In addition, you have written thorough tests for your models and features. That's already enough for me to give you a:

# :+1: 

Great work you guys and you've shown much dedication in learning and excellence. :) Off to your next adventure!

userman95 commented:
thanks Ryan ! You were great aswell! :D See you in next projects

efrapp commented:
Thank you @rvvergara. I learned cool things from your code reviews.
